sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/typed_scrutinee.rs

```diff
@@ -233,55 +235,228 @@ fn type_check_struct(
+                        handler,
+                        field,
+                        has_rest_pattern,
+                        is_public_struct_access,
+                    ) {
+                        Ok(struct_field) => struct_field,
+                        Err(_) => continue,
+                    };
+
+                    // type check the nested scrutinee
+                    let typed_scrutinee = match scrutinee {
+                        None => None,
+                        Some(scrutinee) => Some(ty::TyScrutinee::type_check(
+                            handler,
+                            ctx.by_ref(),
+                            scrutinee.clone(),
+                        )?),
+                    };
+
+                    typed_fields.push(ty::TyStructScrutineeField {
+                        field: field.clone(),
+                        scrutinee: typed_scrutinee,
+                        span: span.clone(),
+                        field_def_name: struct_field.name.clone(),
+                    });
+                }
+                StructScrutineeField::Rest { .. } => {}
             }
         }
-    }
 
-    // ensure that the pattern uses all fields of the struct unless the rest pattern is present
-    if (struct_decl.fields.len() != typed_fields.len()) && rest_pattern.is_none() {
-        let missing_fields = struct_decl
-            .fields
-            .iter()
-            .filter_map(|f| {
-                (!typed_fields.iter().any(|tf| f.name == tf.field)).then_some(f.name.to_string())
-            })
-            .collect::<Vec<_>>();
-
-        return Err(
-            handler.emit_err(CompileError::MatchStructPatternMissingFields {
-                span,
-                missing_fields,
-            }),
-        );
-    }
+        Ok(())
+    })?;
+
+    handler.scope(|handler| {
+        // report struct field privacy errors
+        // This check is intentionally separated from checking the field existence and type-checking the scrutinees.
+        // While we could check private field access immediately after finding the field and emit errors,
+        // that would mean short-circuiting in case of privacy issues which we do not want to do.
+        // The consequence is repeating the search for fields here, but the performance penalty is negligible.
+        if is_public_struct_access {
+            for field in fields {
+                match field {
+                    StructScrutineeField::Field {
+                        field: ref field_name,
+                        ..
+                    } => {
+                        let struct_field = struct_decl
+                            .find_field(field_name)
+                            .expect("The struct field with the given field name must exist.");
+
+                        if struct_field.is_private() {
+                            // TODO: Uncomment this code and delete the one with warnings once struct field privacy becomes a hard error.
```

Comment:
We should create an issue for changing this and mention the issue number in the relevant comments.